pinot-controller/src/main/java/com/linkedin/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -0,0 +1,195 @@
+/**
+ * Copyright (C) 2014-2018 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.api.upload;
+
+import com.linkedin.pinot.common.config.TableNameBuilder;
+import com.linkedin.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadataCustomMapModifier;
+import com.linkedin.pinot.common.metrics.ControllerMeter;
+import com.linkedin.pinot.common.metrics.ControllerMetrics;
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import com.linkedin.pinot.common.utils.FileUploadDownloadClient;
+import com.linkedin.pinot.controller.ControllerConf;
+import com.linkedin.pinot.controller.api.resources.ControllerApplicationException;
+import com.linkedin.pinot.controller.api.resources.FileUploadPathProvider;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.filesystem.PinotFS;
+import com.linkedin.pinot.filesystem.PinotFSFactory;
+import java.io.File;
+import java.net.URI;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import org.apache.helix.ZNRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The ZKOperator is a util class that is used during segment upload to set relevant metadata fields in zk. It will currently
+ * also perform the data move. In the future when we introduce versioning, we will decouple these two steps.
+ */
+public class ZKOperator {
```

Comment:
Breaking this logic into a separate class seems to be the right direction. Good job!

Perhaps we can consider sub-classing this to handle different variations, not sure.

Reply:
I currently do not anticipate any other variations to ZK operations during upload but am open to discussing if you feel there is a need.

Reply:
Will be good if we can add some unit tests to cover the various conditions that this class handles. Not all of it can be covered in integration tests, I believe.

Reply:
I think this should be done in integration tests so that we can verify data as well on the server side. We have integration tests to cover the various conditions (different refresh cases in UploadRefreshIntegrationTest, etc)